tools/crd-generator/crd-generator.go

```diff
@@ -0,0 +1,148 @@
+		Version: v1.OfflineVirtualMachineGroupVersionKind.Version,
+		Scope:   "Namespaced",
+
+		Names: extensionsv1.CustomResourceDefinitionNames{
+			Plural:     "offlinevirtualmachines",
+			Singular:   "offlinevirtualmachine",
+			Kind:       v1.OfflineVirtualMachineGroupVersionKind.Kind,
+			ShortNames: []string{"ovm", "ovms"},
+		},
+		Validation: crdutils.GetCustomResourceValidation("kubevirt.io/kubevirt/pkg/api/v1.OfflineVirtualMachine", v1.GetOpenAPIDefinitions),
+	}
+
+	crdutils.MarshallCrd(crd, "yaml")
+}
+
+func generatePresetCrd() {
+	crd := generateBlankCrd()
+
+	crd.ObjectMeta.Name = "virtualmachinepresets." + v1.VirtualMachinePresetGroupVersionKind.Group
+	crd.Spec = extensionsv1.CustomResourceDefinitionSpec{
+		Group:   v1.VirtualMachinePresetGroupVersionKind.Group,
+		Version: v1.VirtualMachinePresetGroupVersionKind.Version,
+		Scope:   "Namespaced",
+
+		Names: extensionsv1.CustomResourceDefinitionNames{
+			Plural:     "virtualmachinepresets",
+			Singular:   "virtualmachinepreset",
+			Kind:       v1.VirtualMachinePresetGroupVersionKind.Kind,
+			ShortNames: []string{"vmpreset", "vmpresets"},
+		},
+		Validation: crdutils.GetCustomResourceValidation("kubevirt.io/kubevirt/pkg/api/v1.VirtualMachinePreset", v1.GetOpenAPIDefinitions),
+	}
+
+	crdutils.MarshallCrd(crd, "yaml")
+}
+
+func generateReplicaSetCrd() {
+	crd := generateBlankCrd()
+
+	crd.ObjectMeta.Name = "virtualmachinereplicasets." + v1.VMReplicaSetGroupVersionKind.Group
+	crd.Spec = extensionsv1.CustomResourceDefinitionSpec{
+		Group:   v1.VMReplicaSetGroupVersionKind.Group,
+		Version: v1.VMReplicaSetGroupVersionKind.Version,
+		Scope:   "Namespaced",
+
+		Names: extensionsv1.CustomResourceDefinitionNames{
+			Plural:     "virtualmachinereplicasets",
+			Singular:   "virtualmachinereplicaset",
+			Kind:       v1.VMReplicaSetGroupVersionKind.Kind,
+			ShortNames: []string{"vmrs", "vmrss"},
+		},
+		Validation: crdutils.GetCustomResourceValidation("kubevirt.io/kubevirt/pkg/api/v1.VirtualMachineReplicaSet", v1.GetOpenAPIDefinitions),
+	}
+
+	crdutils.MarshallCrd(crd, "yaml")
+}
+
+func generateVirtualMachineCrd() {
+	crd := generateBlankCrd()
+
+	crd.ObjectMeta.Name = "virtualmachines." + v1.VirtualMachineGroupVersionKind.Group
+	crd.Spec = extensionsv1.CustomResourceDefinitionSpec{
+		Group:   v1.VirtualMachineGroupVersionKind.Group,
+		Version: v1.VirtualMachineGroupVersionKind.Version,
+		Scope:   "Namespaced",
+
+		Names: extensionsv1.CustomResourceDefinitionNames{
+			Plural:     "virtualmachines",
+			Singular:   "virtualmachine",
+			Kind:       v1.VirtualMachineGroupVersionKind.Kind,
+			ShortNames: []string{"vm", "vms"},
+		},
+		Validation: crdutils.GetCustomResourceValidation("kubevirt.io/kubevirt/pkg/api/v1.VirtualMachine", v1.GetOpenAPIDefinitions),
+	}
+
+	crdutils.MarshallCrd(crd, "yaml")
+}
+
+func main() {
+	crdType := flag.String("crd-type", "", "Type of crd to generate. vm | vmpreset | vmrs | ovm")
```

Comment:
Is creating it expensive? Is it worth allowing to differentiate them?

Reply:
nah, it's really cheap actually. We have all the openapi parts generated before hand. This function just assembles the openapi v3 schema for a specific object and spits it out to stdout.